src/Streamly/Internal/Data/Stream/StreamD.hs

```diff
@@ -1786,193 +1786,294 @@ data SplitOptions = SplitOptions
-                            r <- done acc
-                            return $ Yield r (GO_SHORT_PAT s)
-                        else go1 SPEC wrd' s acc
-                    else go0 SPEC (idx + 1) wrd' s acc
-                Skip s -> go0 SPEC idx wrd s acc
+                            r <- done fs
+                            let next = SplitOnSeqWordInit 0 (0 :: Word) s
+                            skip $ SplitOnSeqYield r next
+                        else skip $ SplitOnSeqWordLoop wrd1 s fs
+                    else go SPEC (idx + 1) wrd1 s
+                Skip s -> go SPEC idx wrd s
                 Stop -> do
-                    acc' <- if idx /= 0
-                            then go2 wrd idx acc
-                            else return acc
-                    done acc' >>= \r -> return $ Yield r GO_DONE
+                    fs <- initial
+                    if idx /= 0
+                    then skip $ SplitOnSeqWordDone idx fs wrd
+                    else do
+                        r <- done fs
+                        skip $ SplitOnSeqYield r SplitOnSeqDone
 
-        {-# INLINE go1 #-}
-        go1 !_ wrd st !acc = do
+    stepOuter gst (SplitOnSeqWordLoop wrd0 st0 fs0) =
+        go SPEC wrd0 st0 fs0
+
+        where
+
+        {-# INLINE go #-}
+        go !_ !wrd !st !fs = do
             res <- step (adaptState gst) st
             case res of
                 Yield x s -> do
-                    let wrd' = addToWord wrd x
-                        old = (mask .&. wrd) `shiftR` (elemBits * (patLen - 1))
-                    acc' <- fstep acc (toEnum $ fromIntegral old)
-                    if wrd' .&. mask == patWord
-                    then done acc' >>= \r -> return $ Yield r (GO_SHORT_PAT s)
-                    else go1 SPEC wrd' s acc'
-                Skip s -> go1 SPEC wrd s acc
-                Stop -> do
-                    acc' <- go2 wrd patLen acc
-                    done acc' >>= \r -> return $ Yield r GO_DONE
-
-        go2 !wrd !n !acc | n > 0 = do
-            let old = (mask .&. wrd) `shiftR` (elemBits * (n - 1))
-            fstep acc (toEnum $ fromIntegral old) >>= go2 wrd (n - 1)
-        go2 _ _ acc = return acc
+                    let wrd1 = addToWord wrd x
+                        old = (wordMask .&. wrd)
+                                `shiftR` (elemBits * (patLen - 1))
+                    fs1 <- fstep fs (toEnum $ fromIntegral old)
+                    if wrd1 .&. wordMask == wordPat
+                    then do
+                        r <- done fs1
+                        let next = SplitOnSeqWordInit 0 (0 :: Word) s
+                        skip $ SplitOnSeqYield r next
+                    else go SPEC wrd1 s fs1
+                Skip s -> go SPEC wrd s fs
+                Stop -> skip $ SplitOnSeqWordDone patLen fs wrd
+
+    -------------------------------
+    -- General Pattern - Karp Rabin
+    -------------------------------
+
+    stepOuter gst (SplitOnSeqKRInit idx st rb rh) = do
```

Comment:
No loopification here?

Reply:
Does not make any difference, I did it and measured.

Reply:
No difference in performance, dint check fusion though.